cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import dataclasses
+import warnings
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import Dict, List, Tuple, Any, Sequence, Callable, Union, Iterable
+
+import cirq
+import networkx as nx
+from matplotlib import pyplot as plt
+
+
+def cache(user_function: Callable[[Any], Any]) -> Any:
+    """Unbounded cache.
```

Comment:
Yikes.  Are we sure about making this unbounded?  Will this get us into any problems if users do unexpected stuff (like try to enumerate all the lines on a grid or something)?

Reply:
:facepalm: yeah I think it would cause a memory leak. Let me think more about it.

Reply:
Ok, I fixed it by doing `object.__setattr__` during `__post_init__`. I initially discounted this strategy because pycharm/mypy didn't recognize the fields. But there's a loophole: since these dataclasses inherit from the `NamedTopology` base class with these attributes declared, all the tooling seems happy with the latest change.